File: apps/gateway-go/internal/domain/event.go
// Package domain contains the core business entities for the gateway.
// This package has zero external dependencies — it must stay that way.
// All other layers depend on these types; this layer depends on nothing.
package domain

import "time"

// EventType classifies the kind of telemetry being ingested.
// It determines which Redis Stream topic the event is published to.
type EventType string

const (
	EventTypeAudit          EventType = "audit"
	EventTypeLog            EventType = "log"
	EventTypeMetric         EventType = "metric"
	EventTypeTrace          EventType = "trace"
	EventTypeSecurity       EventType = "security"
	EventTypeFrontend       EventType = "frontend"
	EventTypeAI             EventType = "ai"
	EventTypeSystem         EventType = "system"
	EventTypeInfrastructure EventType = "infrastructure"
)

// Severity represents the importance level of an event.
type Severity string

const (
	SeverityDebug    Severity = "debug"
	SeverityInfo     Severity = "info"
	SeverityWarn     Severity = "warn"
	SeverityError    Severity = "error"
	SeverityCritical Severity = "critical"
)

// IsValid returns true if the severity is one of the known values.
func (s Severity) IsValid() bool {
	switch s {
	case SeverityDebug, SeverityInfo, SeverityWarn, SeverityError, SeverityCritical:
		return true
	}
	return false
}

// IsValid returns true if the event type is one of the known values.
func (e EventType) IsValid() bool {
	switch e {
	case EventTypeAudit, EventTypeLog, EventTypeMetric, EventTypeTrace,
		EventTypeSecurity, EventTypeFrontend, EventTypeAI,
		EventTypeSystem, EventTypeInfrastructure:
		return true
	}
	return false
}

// StreamTopic maps the event type to the Redis Stream topic it belongs to.
// All events of the same type are published to the same stream so that
// workers can consume by specialization (e.g. the audit worker only reads
// from stream:audit).
func (e EventType) StreamTopic() string {
	switch e {
	case EventTypeSystem, EventTypeInfrastructure:
		return "stream:system"
	default:
		return "stream:" + string(e)
	}
}

// Event is the canonical telemetry event that flows through the gateway.
// It is populated from the SDK request body and then enriched by the gateway
// before being published to Redis Streams.
type Event struct {
	// Routing fields — resolved by the gateway, not trusted from the request body
	OrganizationID string `json:"organization_id"`
	ApplicationID  string `json:"application_id"`
	Environment    string `json:"environment"`

	// Classification
	EventType EventType `json:"event_type"`
	Severity  Severity  `json:"severity"`

	// Content
	Message string `json:"message"`

	// Timestamps — Timestamp is SDK-reported, IngestedAt is set by the gateway
	Timestamp  time.Time `json:"timestamp"`
	IngestedAt time.Time `json:"ingested_at"`

	// Distributed tracing — optional, populated for trace events
	TraceID      string `json:"trace_id,omitempty"`
	SpanID       string `json:"span_id,omitempty"`
	ParentSpanID string `json:"parent_span_id,omitempty"`

	// Actor — who triggered the event
	UserID    string `json:"user_id,omitempty"`
	SessionID string `json:"session_id,omitempty"`

	// Arbitrary event-specific data — stored as a raw JSON string in ClickHouse
	Payload map[string]any `json:"payload,omitempty"`

	// Source identifies where the event originated.
	// "client" for events submitted via a public token (browser, mobile, desktop).
	// "server" for events submitted via a secret API key from a server-side SDK.
	// Set by the gateway based on the key type — never trusted from SDK input.
	Source string `json:"source,omitempty"`

	// ServiceName is an optional developer-set label identifying the component
	// that sent the event (e.g. "payment-api", "ai-agent", "mobile-ios").
	// Sent by the SDK as X-Service-Name and stored verbatim — not validated.
	ServiceName string `json:"service_name,omitempty"`

	// Gateway-added metadata — set during enrichment, not from SDK
	IPAddress  string `json:"ip_address,omitempty"`
	SDKVersion string `json:"sdk_version,omitempty"`
	Runtime    string `json:"runtime,omitempty"`
	Region     string `json:"region,omitempty"`
}

File: apps/analytics-python/src/domain/event.py
"""
Domain layer — Event entity.

This module defines the canonical Event model used throughout the worker.
It mirrors the Go gateway's domain.Event struct so both services speak the
same language when passing events through Redis Streams.

Zero external I/O — only pydantic for validation. No Redis, no ClickHouse.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any
import json

from pydantic import BaseModel, Field, field_validator


class EventType(str, Enum):
    """
    Classifies the kind of telemetry event.
    Maps directly to the Redis Stream topic and ClickHouse event_type column.
    """
    AUDIT = "audit"
    LOG = "log"
    METRIC = "metric"
    TRACE = "trace"
    SECURITY = "security"
    FRONTEND = "frontend"
    AI = "ai"
    SYSTEM = "system"
    INFRASTRUCTURE = "infrastructure"


class Severity(str, Enum):
    """
    Importance level of an event, from least to most critical.
    Stored as a LowCardinality(String) in ClickHouse for fast filtering.
    """
    DEBUG = "debug"
    INFO = "info"
    WARN = "warn"
    ERROR = "error"
    CRITICAL = "critical"


class Event(BaseModel):
    """
    The canonical telemetry event flowing through the worker pipeline.

    Events arrive as flat string maps from Redis Streams (published by the
    Go Gateway) and are parsed into this model before being written to
    ClickHouse. All fields match the ClickHouse `watcher.events` schema.
    """

    # Routing — who owns this event
    organization_id: str
    application_id: str = ""
    environment: str = ""

    # Classification
    event_type: EventType
    severity: Severity

    # Content
    message: str

    # Source — set by the gateway based on the API key type, never by the SDK.
    # "client" = public token (wpub_) used from browser, mobile, or desktop.
    # "server" = secret key (wt_) used from a server-side SDK.
    # ""       = legacy events ingested before source tagging was introduced.
    source: str = ""

    # ServiceName is an optional developer-set label for the sending component.
    # Set at SDK init time (e.g. "payment-api", "ai-agent", "mobile-ios").
    # Sent as X-Service-Name header — stored verbatim, never validated.
    service_name: str = ""

    # Timestamps — both always present, ingested_at is server-set by gateway
    timestamp: datetime
    ingested_at: datetime

    # Distributed tracing — optional
    trace_id: str = ""
    span_id: str = ""
    parent_span_id: str = ""

    # Actor
    user_id: str = ""
    session_id: str = ""

    # Arbitrary event-specific data — stored as JSON string in ClickHouse
    payload: str = Field(default="{}")

    # Gateway-added metadata
    ip_address: str = ""
    sdk_version: str = ""
    runtime: str = ""
    region: str = ""

    @field_validator("payload", mode="before")
    @classmethod
    def coerce_payload(cls, v: Any) -> str:
        """
        Accepts either a JSON string or a dict and normalizes to a JSON string.
        Redis Streams store everything as strings, but the gateway may also
        pass a dict directly in tests — handle both.
        """
        if isinstance(v, dict):
            return json.dumps(v)
        if v is None or v == "":
            return "{}"
        return str(v)

    @classmethod
    def from_stream_fields(cls, fields: dict[str, str]) -> "Event":
        """
        Deserializes a flat Redis Stream message (all string values) into
        an Event domain object. Redis Streams store everything as bytes/strings,
        so this is the canonical entry point for converting stream data.
        """
        return cls.model_validate(fields)

    def to_clickhouse_row(self) -> tuple:
        """
        Returns the event as a tuple in the exact column order expected by
        the ClickHouse `watcher.events` table INSERT statement.

        Column order must match the INSERT in clickhouse_adapter/repository.py.
        """
        return (
            self.organization_id,
            self.application_id,
            self.environment,
            self.event_type.value,
            self.severity.value,
            self.message,
            self.source,
            self.service_name,
            self.timestamp,
            self.trace_id,
            self.span_id,
            self.parent_span_id,
            self.user_id,
            self.session_id,
            self.payload,
            self.ingested_at,
            self.sdk_version,
            self.runtime,
            self.ip_address,
            self.region,
        )
